File: Apps/Android/app/src/main/kotlin/com/opencapture/openpocketcine/lut/LUTPicker.kt
package com.opencapture.openpocketcine.lut

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.OpenableColumns
import androidx.activity.compose.BackHandler
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.outlined.RadioButtonUnchecked
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.opencapture.openpocketcine.AppModel
import com.opencapture.openpocketcine.LiveDesign
import com.opencapture.openpocketcine.chromeClickable
import com.opencapture.openpocketcine.pairing.StartupColors
import java.io.File

private enum class LutTab(val label: String) {
    BUILT_IN("Built-in"),
    DJI("DJI"),
    CUSTOM("Custom"),
}

@Composable
fun LUTPicker(model: AppModel, onClose: () -> Unit) {
    val context = LocalContext.current
    val djiEntries =
        remember(context) {
            val names = context.assets.list(LutCatalog.ASSET_DIRECTORY)?.toList().orEmpty()
            LutCatalog.djiEntries(names)
        }
    var tab by remember {
        mutableStateOf(
            when (LutCatalog.categoryOf(model.lutSelection)) {
                LutCategory.DJI -> LutTab.DJI
                LutCategory.CUSTOM -> LutTab.CUSTOM
                LutCategory.BUILT_IN -> LutTab.BUILT_IN
            }
        )
    }
    var splitComparison by remember { mutableStateOf(false) }
    var splitVertical by remember { mutableStateOf(true) }
    var customRevision by remember { mutableIntStateOf(0) }
    var pendingDeletion by remember { mutableStateOf<String?>(null) }
    var deletionError by remember { mutableStateOf<String?>(null) }
    var importError by remember { mutableStateOf<String?>(null) }
    val customDir = remember(context) { LutCatalog.customDirectory(context.filesDir) }
    val customEntries = remember(customRevision, customDir) { LutCatalog.listCustom(customDir) }
    val importer =
        rememberLauncherForActivityResult(ActivityResultContracts.OpenDocument()) { uri ->
            if (uri == null) return@rememberLauncherForActivityResult
            runCatching { importCube(context, uri, customDir) }
                .onSuccess { entry ->
                    importError = null
                    customRevision += 1
                    val current = model.lutSelection
                    if (current != LutCatalog.AUTO && current != LutCatalog.DJI_AUTO) {
                        model.updateLutSelection(entry.id)
                    }
                }
                .onFailure { error ->
                    importError = error.message ?: "The .cube file could not be read."
                }
        }

    BackHandler(onBack = onClose)
    Column(
        Modifier.fillMaxSize()
            .background(LiveDesign.background)
            .statusBarsPadding()
            .navigationBarsPadding()
            .padding(20.dp)
    ) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Text("LUT", color = LiveDesign.text, fontSize = 17.sp, fontWeight = FontWeight.Bold)
            Spacer(Modifier.weight(1f))
            Text(
                "Done",
                color = LiveDesign.accent,
                fontSize = 16.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.chromeClickable(onClick = onClose).padding(horizontal = 8.dp, vertical = 4.dp),
            )
        }
        Spacer(Modifier.height(12.dp))
        LutSegmentedButtons(
            items = LutTab.entries.map { it.label },
            selected = tab.label,
        ) { label ->
            val next = LutTab.entries.first { it.label == label }
            if (next == tab) return@LutSegmentedButtons
            tab = next
            when (next) {
                LutTab.BUILT_IN ->
                    if (LutCatalog.categoryOf(model.lutSelection) != LutCategory.BUILT_IN) {
                        model.updateLutSelection(LutCatalog.AUTO)
                    }
                LutTab.DJI ->
                    if (LutCatalog.categoryOf(model.lutSelection) != LutCategory.DJI) {
                        model.updateLutSelection(LutCatalog.DJI_AUTO)
                    }
                LutTab.CUSTOM -> Unit
            }
        }
        Spacer(Modifier.height(8.dp))
        Column(Modifier.weight(1f).fillMaxWidth()) {
            when (tab) {
                LutTab.BUILT_IN ->
                    CatalogTab(
                        caption = builtInCaption(model.lutSelection),
                        entries = LutCatalog.builtIn,
                        selection = model.lutSelection,
                        onSelect = model::updateLutSelection,
                    )
                LutTab.DJI ->
                    CatalogTab(
                        caption = djiCaption(model.lutSelection),
                        entries = djiEntries,
                        selection = model.lutSelection,
                        onSelect = model::updateLutSelection,
                    )
                LutTab.CUSTOM ->
                    CustomTab(
                        imported = customEntries,
                        selection = model.lutSelection,
                        onImport = { importer.launch(arrayOf("*/*")) },
                        onSelect = model::updateLutSelection,
                        onClear = { pendingDeletion = it },
                    )
            }
        }
        Spacer(Modifier.height(8.dp))
        LutSplitComparisonBar(
            splitComparison = splitComparison,
            splitVertical = splitVertical,
            onToggleSplit = { splitComparison = !splitComparison },
            onSplitVertical = { splitVertical = it },
        )
        if (importError != null) {
            Text(
                importError.orEmpty(),
                color = StartupColors.destructive,
                fontSize = 12.sp,
                modifier = Modifier.padding(top = 8.dp),
            )
        }
    }

    val pending = pendingDeletion
    if (pending != null) {
        AlertDialog(
            onDismissRequest = { pendingDeletion = null },
            title = { Text("Clear ${LutCatalog.displayName(pending)}?") },
            text = { Text("This removes the stored LUT from this device. This action cannot be undone.") },
            confirmButton = {
                TextButton(
                    onClick = {
                        pendingDeletion = null
                        runCatching { LutCatalog.deleteCustom(pending, customDir) }
                            .onSuccess {
                                customRevision += 1
                                if (model.lutSelection == LutCatalog.customId(pending)) {
                                    model.updateLutSelection(LutCatalog.AUTO)
                                }
                            }
                            .onFailure { error ->
                                deletionError = error.message ?: "The LUT could not be deleted."
                            }
                    }
                ) { Text("Clear LUT", color = StartupColors.destructive) }
            },
            dismissButton = { TextButton(onClick = { pendingDeletion = null }) { Text("Cancel") } },
        )
    }
    if (deletionError != null) {
        AlertDialog(
            onDismissRequest = { deletionError = null },
            title = { Text("Couldn’t Delete LUT") },
            text = { Text(deletionError ?: "The LUT could not be deleted.") },
            confirmButton = { TextButton(onClick = { deletionError = null }) { Text("OK") } },
        )
    }
}

@Composable
private fun CatalogTab(
    caption: String,
    entries: List<LutEntry>,
    selection: String,
    onSelect: (String) -> Unit,
) {
    Column(Modifier.fillMaxSize()) {
        Text(
            caption,
            color = LiveDesign.muted,
            fontSize = 11.sp,
            fontWeight = FontWeight.Medium,
            modifier = Modifier.fillMaxWidth().padding(bottom = 4.dp),
        )
        Column(Modifier.weight(1f).fillMaxWidth().verticalScroll(rememberScrollState())) {
            entries.forEach { entry ->
                val selected = LutCatalog.matches(entry, selection)
                Text(
                    entry.title,
                    color = if (selected) LiveDesign.accent else LiveDesign.muted.copy(alpha = 0.7f),
                    fontSize = if (selected) 22.sp else 16.sp,
                    fontWeight = if (selected) FontWeight.SemiBold else FontWeight.Normal,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier =
                        Modifier.fillMaxWidth()
                            .clip(RoundedCornerShape(12.dp))
                            .background(if (selected) LiveDesign.accentDim else LiveDesign.glassBright)
                            .clickable { onSelect(entry.id) }
                            .padding(horizontal = 12.dp, vertical = 14.dp),
                )
                Spacer(Modifier.height(6.dp))
            }
        }
    }
}

@Composable
private fun CustomTab(
    imported: List<LutEntry>,
    selection: String,
    onImport: () -> Unit,
    onSelect: (String) -> Unit,
    onClear: (String) -> Unit,
) {
    Column(Modifier.fillMaxSize()) {
        Box(
            Modifier.fillMaxWidth()
                .clip(RoundedCornerShape(50))
                .background(LiveDesign.glassBright)
                .clickable(onClick = onImport)
                .padding(vertical = 8.dp),
            contentAlignment = Alignment.Center,
        ) {
            Text("Import .cube", color = LiveDesign.text, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
        }
        Spacer(Modifier.height(6.dp))
        if (imported.isEmpty()) {
            Box(Modifier.weight(1f).fillMaxWidth(), contentAlignment = Alignment.Center) {
                Text(
                    "Imported looks land here. Auto stays on Built-in or DJI.",
                    color = LiveDesign.muted,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Medium,
                )
            }
        } else {
            Column(Modifier.weight(1f).fillMaxWidth().verticalScroll(rememberScrollState())) {
                imported.forEach { entry ->
                    val selected = LutCatalog.matches(entry, selection)
                    val fileName = entry.fileName ?: return@forEach
                    Row(
                        Modifier.fillMaxWidth()
                            .clip(RoundedCornerShape(LiveDesign.CORNER_RADIUS_DP.dp))
                            .background(if (selected) LiveDesign.accentDim else LiveDesign.glassBright)
                            .padding(horizontal = 10.dp, vertical = 6.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(6.dp),
                    ) {
                        Text(
                            entry.title,
                            color = if (selected) LiveDesign.accent else LiveDesign.text,
                            fontSize = 13.sp,
                            fontWeight = FontWeight.SemiBold,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                            modifier =
                                Modifier.weight(1f)
                                    .clickable { onSelect(entry.id) }
                                    .padding(vertical = 6.dp),
                        )
                        Box(
                            Modifier.clip(RoundedCornerShape(50))
                                .background(LiveDesign.glassBright)
                                .clickable { onClear(fileName) }
                                .padding(horizontal = 8.dp, vertical = 6.dp),
                        ) {
                            Text("Clear", color = StartupColors.destructive, fontSize = 12.sp, fontWeight = FontWeight.SemiBold)
                        }
                    }
                    Spacer(Modifier.height(6.dp))
                }
            }
        }
    }
}

@Composable
private fun LutSplitComparisonBar(
    splitComparison: Boolean,
    splitVertical: Boolean,
    onToggleSplit: () -> Unit,
    onSplitVertical: (Boolean) -> Unit,
) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
        Row(
            Modifier.clip(RoundedCornerShape(50))
                .background(if (splitComparison) LiveDesign.accentDim else LiveDesign.glassBright)
                .clickable(onClick = onToggleSplit)
                .padding(horizontal = 12.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Icon(
                imageVector = if (splitComparison) Icons.Filled.CheckCircle else Icons.Outlined.RadioButtonUnchecked,
                contentDescription = null,
                tint = if (splitComparison) LiveDesign.accent else LiveDesign.muted,
                modifier = Modifier.size(18.dp),
            )
            Text("50/50", color = LiveDesign.text, fontSize = 14.sp, fontWeight = FontWeight.SemiBold)
        }
        if (splitComparison) {
            LutSegmentedButtons(
                items = listOf("Left / Right", "Top / Bottom"),
                selected = if (splitVertical) "Left / Right" else "Top / Bottom",
                modifier = Modifier.weight(1f),
            ) { label ->
                onSplitVertical(label == "Left / Right")
            }
        } else {
            Spacer(Modifier.weight(1f))
        }
    }
}

@Composable
private fun LutSegmentedButtons(
    items: List<String>,
    selected: String,
    modifier: Modifier = Modifier,
    onSelect: (String) -> Unit,
) {
    Row(modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        items.forEach { item ->
            val on = item == selected
            Box(
                Modifier.weight(1f)
                    .clip(RoundedCornerShape(50))
                    .background(if (on) LiveDesign.accentDim else LiveDesign.glassBright)
                    .clickable { onSelect(item) }
                    .padding(vertical = 10.dp),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    item,
                    color = if (on) LiveDesign.accent else LiveDesign.muted,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                )
            }
        }
    }
}

private fun builtInCaption(selection: String): String =
    when (selection) {
        LutCatalog.AUTO -> "App looks for this color / camera"
        LutCatalog.OFF -> "No matching look for this color / camera"
        else -> "App looks for this color / camera"
    }

private fun djiCaption(selection: String): String =
    when (selection) {
        LutCatalog.DJI_AUTO -> "Official DJI looks for this color / camera"
        else -> "Official DJI Rec.709 cube"
    }

private fun importCube(context: Context, uri: Uri, directory: File): LutEntry {
    runCatching {
        context.contentResolver.takePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
    }
    val rawName = queryDisplayName(context, uri) ?: uri.lastPathSegment ?: "Imported.cube"
    val bytes =
        context.contentResolver.openInputStream(uri)?.use { it.readBytes() }
            ?: throw IllegalStateException("The .cube file could not be read.")
    return LutCatalog.importCube(rawName, bytes, directory)
}

private fun queryDisplayName(context: Context, uri: Uri): String? {
    val cursor = context.contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)
    cursor?.use {
        if (it.moveToFirst()) {
            val index = it.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            if (index >= 0) return it.getString(index)
        }
    }
    return null
}

File: ios/OpenPocketCine/LUTPicker.swift
import OpenPocketViewCore
import SwiftUI
import UniformTypeIdentifiers

/// Built-in / DJI / Custom. 50/50 is pinned on the long-press footer so
/// landscape never hides it under the catalog; the sheet keeps it inline.
/// `embedded` is the assist-tray form; the sheet wraps the same body.
struct LUTPicker: View {
    private enum Category: String, CaseIterable {
        case builtIn = "Built-in"
        case dji = "DJI"
        case custom = "Custom"
    }

    @Bindable var assist: LiveAssistState
    var embedded = false
    @Environment(\.dismiss) private var dismiss
    @Environment(AppModel.self) private var model
    @State private var category: Category = .builtIn
    @State private var importing = false
    @State private var pendingDeletion: String?
    @State private var deletionErrorMessage: String?
    @State private var importError: String?

    /// OpenZCine `LUTPickerContent.contentHeight` — landscape has no headroom for a taller drum.
    private let contentHeight: CGFloat = 146
    private let captionHeight: CGFloat = 28

    var body: some View {
        if embedded {
            picker
        } else {
            NavigationStack {
                picker
                    .padding(20)
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
                    .background(LiveDesign.background.ignoresSafeArea())
                    .navigationTitle("LUT")
                    .navigationBarTitleDisplayMode(.inline)
                    .toolbar {
                        ToolbarItem(placement: .cancellationAction) {
                            Button("Done") { dismiss() }
                                .foregroundStyle(LiveDesign.accent)
                        }
                    }
            }
            .preferredColorScheme(.dark)
        }
    }

    private var picker: some View {
        VStack(spacing: 8) {
            LUTSegmentedButtons(
                items: Category.allCases.map(\.rawValue),
                selected: category.rawValue
            ) { raw in
                guard let next = Category(rawValue: raw), next != category else { return }
                category = next
                switch next {
                case .builtIn:
                    if !assist.lutSelection.isBuiltIn { assist.selectLUT(.auto) }
                case .dji:
                    if !assist.lutSelection.isDJI { assist.selectLUT(.djiAuto) }
                case .custom:
                    break
                }
            }
            tabContent
                .frame(maxWidth: .infinity)
                .frame(height: contentHeight)
            // Embedded long-press parks 50/50 on the panel footer so it
            // stays on screen when the catalog scrolls.
            if !embedded {
                LUTSplitComparisonBar(assist: assist)
            }
            if let importError {
                Text(importError)
                    .font(LiveType.ui(size: 12, weight: .regular, design: .rounded))
                    .foregroundStyle(StartupColors.destructive)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .onAppear {
            if assist.lutSelection.isCustom {
                category = .custom
            } else if assist.lutSelection.isDJI {
                category = .dji
            } else {
                category = .builtIn
            }
            assist.syncLUT(
                to: model.session.status.colorMode,
                family: model.session.bodyFamily,
                cameraName: model.session.connectedCamera?.model.name)
        }
        .confirmationDialog(
            pendingDeletion.map { "Clear \(CustomLUTIndex.displayName(fileName: $0))?" }
                ?? "Clear LUT?",
            isPresented: Binding(
                get: { pendingDeletion != nil },
                set: { if !$0 { pendingDeletion = nil } }),
            titleVisibility: .visible
        ) {
            if let pendingDeletion {
                Button("Clear LUT", role: .destructive) {
                    clear(pendingDeletion)
                }
            }
            Button("Cancel", role: .cancel) { pendingDeletion = nil }
        } message: {
            Text("This removes the stored LUT from this device. This action cannot be undone.")
        }
        .alert(
            "Couldn’t Delete LUT",
            isPresented: Binding(
                get: { deletionErrorMessage != nil },
                set: { if !$0 { deletionErrorMessage = nil } })
        ) {
            Button("OK") { deletionErrorMessage = nil }
        } message: {
            Text(deletionErrorMessage ?? "The LUT could not be deleted.")
        }
        .fileImporter(
            isPresented: $importing,
            allowedContentTypes: [UTType(filenameExtension: "cube") ?? .data]
        ) { result in
            handleImport(result)
        }
    }

    @ViewBuilder private var tabContent: some View {
        switch category {
        case .builtIn:
            catalogTab(
                cases: LUTSelection.builtInCases,
                inCatalog: assist.lutSelection.isBuiltIn,
                fallback: .auto,
                caption: builtInCaption
            )
        case .dji:
            catalogTab(
                cases: LUTSelection.djiCases,
                inCatalog: assist.lutSelection.isDJI,
                fallback: .djiAuto,
                caption: djiCaption
            )
        case .custom:
            customTab
        }
    }

    private func catalogTab(
        cases: [LUTSelection],
        inCatalog: Bool,
        fallback: LUTSelection,
        caption: String
    ) -> some View {
        VStack(spacing: 4) {
            Text(caption)
                .font(LiveType.ui(size: 11, weight: .medium, design: .rounded))
                .foregroundStyle(LiveDesign.muted)
                .multilineTextAlignment(.center)
                .frame(height: captionHeight)
            LUTDrumWheel(
                options: cases.map(\.title),
                selection: Binding(
                    get: {
                        inCatalog ? assist.lutSelection.title : fallback.title
                    },
                    set: { name in
                        guard let selection = cases.first(where: { $0.title == name }) else {
                            return
                        }
                        assist.selectLUT(selection)
                    }
                ),
                wheelHeight: contentHeight - captionHeight - 4
            )
        }
    }

    private var builtInCaption: String {
        if assist.lutSelection == .auto {
            return LUTResolver.autoCaption(source: assist.resolvedSource())
        }
        if assist.lutSelection == .officialDLog {
            return "Built-in D-Log look"
        }
        if assist.lutSelection == .officialDLog2 {
            return "Built-in D-Log2 look"
        }
        return "App looks for this color / camera"
    }

    private var djiCaption: String {
        if assist.lutSelection == .djiAuto {
            return LUTResolver.autoCaption(source: assist.resolvedSource())
        }
        if assist.lutSelection.isDJI {
            return "Official DJI Rec.709 cube"
        }
        return "Official DJI looks for this color / camera"
    }

    @ViewBuilder private var customTab: some View {
        let imported = CustomLUTStore.list()
        VStack(spacing: 6) {
            Button {
                importing = true
            } label: {
                Text("Import .cube")
                    .font(LiveType.ui(size: 13, weight: .semibold, design: .rounded))
                    .foregroundStyle(LiveDesign.text)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 8)
                    .background(LiveDesign.glassBright, in: Capsule())
            }
            .buttonStyle(.zcTapTarget)
            if imported.isEmpty {
                Text("Imported looks land here. Auto stays on Built-in or DJI.")
                    .font(LiveType.ui(size: 11, weight: .medium, design: .rounded))
                    .foregroundStyle(LiveDesign.muted)
                    .multilineTextAlignment(.center)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                ScrollView {
                    VStack(spacing: 6) {
                        ForEach(imported, id: \.fileName) { lut in
                            customFileRow(lut)
                        }
                    }
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
    }

    private func customFileRow(_ lut: StoredCustomLUT) -> some View {
        let selected =
            assist.lutSelection == .customFile
            && OperatorPrefs.selectedCustomFileName == lut.fileName
        return HStack(spacing: 6) {
            Button {
                assist.selectCustomFile(lut.fileName)
            } label: {
                Text(lut.displayName)
                    .font(LiveType.ui(size: 13, weight: .semibold, design: .rounded))
                    .foregroundStyle(selected ? LiveDesign.accent : LiveDesign.text)
                    .lineLimit(1)
                    .minimumScaleFactor(0.75)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.zcTapTarget)
            Button {
                pendingDeletion = lut.fileName
            } label: {
                Text("Clear")
                    .font(LiveType.ui(size: 12, weight: .semibold, design: .rounded))
                    .foregroundStyle(StartupColors.destructive)
                    .padding(.vertical, 6)
                    .padding(.horizontal, 8)
                    .background(LiveDesign.glassBright, in: Capsule())
            }
            .buttonStyle(.zcTapTarget)
            .fixedSize()
        }
        .padding(.vertical, 6)
        .padding(.horizontal, 10)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            selected ? LiveDesign.accentDim : LiveDesign.glassBright,
            in: RoundedRectangle(cornerRadius: LiveDesign.cornerRadius, style: .continuous)
        )
    }

    private func handleImport(_ result: Result<URL, Error>) {
        switch result {
        case .success(let url):
            importCube(url)
        case .failure(let error):
            importError = error.localizedDescription
        }
    }

    private func importCube(_ url: URL) {
        let accessed = url.startAccessingSecurityScopedResource()
        defer { if accessed { url.stopAccessingSecurityScopedResource() } }
        do {
            let stored = try CustomLUTStore.importFile(from: url)
            guard CustomLUTStore.cube(stored) != nil else {
                try? CustomLUTStore.remove(stored)
                throw CustomLUTStoreError.unreadable
            }
            assist.importCustom(fileName: stored.fileName)
            importError = nil
        } catch {
            importError = error.localizedDescription
        }
    }

    private func clear(_ fileName: String) {
        pendingDeletion = nil
        assist.clearCustomFile(fileName)
    }
}

/// OpenZCine 50/50: off-by-default toggle; orientation chips only while armed.
/// Labels are `Left / Right` and `Top / Bottom`, not Vertical/Horizontal.
///
/// The long-press panel renders this as a pinned footer so opening LUT never
/// hides the control under a catalog scroll.
struct LUTSplitComparisonBar: View {
    @Bindable var assist: LiveAssistState

    var body: some View {
        HStack(spacing: 10) {
            Button {
                assist.splitComparison.toggle()
                if assist.splitComparison {
                    if !assist.lutArmed { assist.armLastLUT() }
                }
                assist.persist()
            } label: {
                HStack(spacing: 8) {
                    (assist.splitComparison ? OpcIcon.circleCheck : OpcIcon.circle)
                        .frame(width: 16, height: 16)
                        .foregroundStyle(
                            assist.splitComparison ? LiveDesign.accent : LiveDesign.muted)
                    Text("50/50")
                        .font(LiveType.ui(size: 14, weight: .semibold, design: .rounded))
                        .foregroundStyle(LiveDesign.text)
                }
                .padding(.vertical, 8)
                .padding(.horizontal, 12)
                .background(
                    assist.splitComparison
                        ? LiveDesign.accentDim : LiveDesign.glassBright,
                    in: Capsule())
            }
            .buttonStyle(.zcTapTarget)
            if assist.splitComparison {
                LUTSegmentedButtons(
                    items: LUTSplitOrientation.allCases.map(\.rawValue),
                    selected: LUTSplitOrientation.current(vertical: assist.splitVertical).rawValue
                ) { raw in
                    guard let next = LUTSplitOrientation(rawValue: raw) else { return }
                    assist.splitVertical = next.isVertical
                    assist.persist()
                }
            } else {
                Spacer(minLength: 0)
            }
        }
    }
}

/// OpenZCine `SplitComparisonOrientation` labels.
private enum LUTSplitOrientation: String, CaseIterable {
    case leftRight = "Left / Right"
    case topBottom = "Top / Bottom"

    var isVertical: Bool { self == .leftRight }

    static func current(vertical: Bool) -> LUTSplitOrientation {
        vertical ? .leftRight : .topBottom
    }
}

/// OpenZCine `SegmentedButtons` — capsule chips, gold when selected.
private struct LUTSegmentedButtons: View {
    let items: [String]
    let selected: String
    let onSelect: (String) -> Void

    var body: some View {
        HStack(spacing: 6) {
            ForEach(items, id: \.self) { item in
                Button {
                    onSelect(item)
                } label: {
                    Text(item)
                        .font(LiveType.ui(size: 13, weight: .semibold, design: .rounded))
                        .foregroundStyle(item == selected ? LiveDesign.accent : LiveDesign.muted)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 10)
                        .background(
                            item == selected ? LiveDesign.accentDim : LiveDesign.glassBright,
                            in: Capsule()
                        )
                }
                .buttonStyle(.zcTapTarget)
            }
        }
    }
}

/// OpenZCine `AccentDrumWheel` subset used by the LUT picker (height + delete).
private struct LUTDrumWheel: View {
    let options: [String]
    @Binding var selection: String
    var wheelHeight: CGFloat = 146
    var onDeleteOption: ((String) -> Void)? = nil

    private let rowHeight: CGFloat = 52

    var body: some View {
        ScrollViewReader { proxy in
            ScrollView(.vertical, showsIndicators: false) {
                VStack(spacing: 0) {
                    ForEach(options, id: \.self) { option in
                        optionRow(option, isCentered: option == selection)
                    }
                }
                .scrollTargetLayout()
            }
            .scrollTargetBehavior(.viewAligned)
            .scrollPosition(
                id: Binding(get: { selection }, set: { if let value = $0 { selection = value } })
            )
            .contentMargins(.vertical, max(0, (wheelHeight - rowHeight) / 2), for: .scrollContent)
            .frame(height: wheelHeight)
            .sensoryFeedback(.selection, trigger: selection)
            .mask {
                LinearGradient(
                    stops: [
                        .init(color: .clear, location: 0),
                        .init(color: .black, location: 0.22),
                        .init(color: .black, location: 0.78),
                        .init(color: .clear, location: 1),
                    ],
                    startPoint: .top,
                    endPoint: .bottom
                )
            }
            .overlay {
                Rectangle().fill(LiveDesign.hairlineStrong).frame(height: 1)
                    .offset(y: -rowHeight / 2)
                Rectangle().fill(LiveDesign.hairlineStrong).frame(height: 1)
                    .offset(y: rowHeight / 2)
            }
            .onAppear {
                DispatchQueue.main.async { proxy.scrollTo(selection, anchor: .center) }
            }
        }
    }

    @ViewBuilder private func optionRow(_ option: String, isCentered: Bool) -> some View {
        let row = Text(option)
            .font(
                .system(
                    size: isCentered ? 30 : 23,
                    weight: isCentered ? .semibold : .regular,
                    design: .monospaced)
            )
            .lineLimit(1)
            .minimumScaleFactor(0.5)
            .foregroundStyle(isCentered ? LiveDesign.accent : LiveDesign.muted.opacity(0.7))
            .frame(maxWidth: .infinity)
            .frame(height: rowHeight)
            .contentShape(Rectangle())
            .id(option)

        if let onDeleteOption {
            row
                .contextMenu {
                    Button(role: .destructive) {
                        onDeleteOption(option)
                    } label: {
                        Label {
                            Text("Delete LUT")
                        } icon: {
                            OpcIcon.trash
                        }
                    }
                }
                .accessibilityAction(named: Text("Delete \(option)")) {
                    onDeleteOption(option)
                }
        } else {
            row
        }
    }
}
